alibi/explainers/similarity/backends/pytorch/base.py

```diff
@@ -50,8 +50,26 @@ def get_grads(
         loss = loss_fn(output, Y)
         loss.backward()
         model.train(initial_model_state)
-        return np.concatenate([_PytorchBackend.to_numpy(param.grad).reshape(-1)  # type: ignore [arg-type] # see #810
-                               for param in model.parameters()])
+
+        return np.concatenate([_PytorchBackend._grad_to_numpy(grad=param.grad, name=name)
+                               for name, param in model.named_parameters()
+                               if param.grad is not None])
+
+    @staticmethod
+    def _grad_to_numpy(grad: torch.Tensor, name: Optional[str] = None) -> torch.Tensor:
+        """Convert graidient to numpy array.
+
+        Converts gradient tensor to flat numpy array. If the gradient is a sparse tensor, it is converted to a dense \
+        tensor first.
+        """
+        if grad.is_sparse:
+            grad = grad.to_dense()
+
+        if not hasattr(grad, 'numpy'):
+            name = f' for the named tensor: {name}' if name else ''
+            raise TypeError((f'Could not convert gradient to numpy array{name}. To ignore these '
+                             'gradients in the similarity computation use `requires_grad=False`.'))
```

Comment:
Is it possible for the user to easily do this? I'm thinking of the situation where a layer has multiple gradient tensors, e.g. weights and bias. Here we would be reporting the tensors not the layers?

A related point might be that this could be annoying if many such layers are present, as the error would be thrown for the first one, and then once it's fixed the next one etc. Should we consider checking the `numpy` attribute of all tensors first (after converting to dense) and then raise one error if there are any that don't have that attribute? Perhaps this is a minor point if in realistic scenarios we don't expect there to be more than a handful of such layers?

Reply:
So it sounds like this check is more of a fail-safe and shouldn't really happen unless something is quite wrong? In which case I think it's fine as we at least have a readable error rather than a non-readable one.